protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -282,32 +274,24 @@ impl NetworkBehaviour for Behaviour {
         addr: &Multiaddr,
         role_override: Endpoint,
     ) -> Result<THandler<Self>, ConnectionDenied> {
-        match self
-            .outgoing_direct_connection_attempts
-            .remove(&(connection_id, peer))
+        if is_relayed(addr) {
+            return Ok(Either::Left(handler::relayed::Handler::new(
+                ConnectedPoint::Dialer {
+                    address: addr.clone(),
+                    role_override,
+                },
+            ))); // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
+        }
+
+        if let Some(&relayed_connection_id) = self.direct_to_relayed_connections.get(&connection_id)
         {
-            None => {
-                let handler = if is_relayed(addr) {
-                    Either::Left(handler::relayed::Handler::new(ConnectedPoint::Dialer {
-                        address: addr.clone(),
-                        role_override,
-                    })) // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
-                } else {
-                    Either::Right(Either::Right(dummy::ConnectionHandler))
-                };
-
-                Ok(handler)
-            }
-            Some(_) => {
-                assert!(
-                    !is_relayed(addr),
-                    "`Prototype::DirectConnection` is never created for relayed connection."
-                );
-
-                Ok(Either::Right(Either::Left(
-                    handler::direct::Handler::default(),
-                )))
-            }
```

Comment:
Am I confusing myself now or wasn't this completely broken? If `outgoing_direct_connection_attempts` is indexed by relayed connection IDs, shouldn't we have hit this assert for a failed holepunch? (Which has an outdated message :see_no_evil: ).

Reply:
Like I mentioned before, the `Some` arm wouldn't have ever executed. `outgoing_direct_connection_attempts` gets an entry only *after* the direct connection negotiation (over the relay) completes.

Reply:
True, because the connection ID for the relay connection can never appear here again. Any new connection would get a new ID.